tarpc/src/transport/mod.rs

```diff
@@ -0,0 +1,84 @@
+use std::io::{self, Read, Write};
+use std::time::Duration;
+
+/// A factory for creating a listener on a given address.
+pub trait Transport {
+    /// The type of listener that binds to the given address.
+    type Listener: Listener;
+    /// Return a listener on the given address, and a dialer to that address.
+    fn bind(&self) -> io::Result<Self::Listener>;
+}
+
+/// Accepts incoming connections from dialers.
+pub trait Listener: Send + 'static {
+    /// The type of address being listened on.
+    type Dialer: Dialer;
+    /// The type of stream this listener accepts.
+    type Stream: Stream;
+    /// Accept an incoming stream.
+    fn accept(&self) -> io::Result<Self::Stream>;
+    /// Returns the local address being listened on.
+    fn dialer(&self) -> io::Result<Self::Dialer>;
+    /// Iterate over incoming connections.
+    fn incoming(&self) -> Incoming<Self> {
+        Incoming {
+            listener: self,
+        }
+    }
+}
+
+/// A cloneable Reader/Writer.
+pub trait Stream: Read + Write + Send + Sized + 'static {
+    /// Clone that can fail.
+    fn try_clone(&self) -> io::Result<Self>;
+    /// Sets a read timeout.
+    fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()>;
```

Comment:
It was just because TcpStrean's equivalent timeout methods take an Option.
On Feb 24, 2016 7:47 PM, "shaladdle" <notifications@github.com> wrote:

> In tarpc/src/transport/mod.rs
> <https://github.com/google/tarpc/pull/27#discussion_r54045594>:
>
> > +    /// Returns the local address being listened on.
> > +    fn dialer(&self) -> io::Result<Self::Dialer>;
> > +    /// Iterate over incoming connections.
> > +    fn incoming(&self) -> Incoming<Self> {
> > +        Incoming {
> > +            listener: self,
> > +        }
> > +    }
> > +}
> > +
> > +/// A cloneable Reader/Writer.
> > +pub trait Stream: Read + Write + Send + Sized + 'static {
> > +    /// Clone that can fail.
> > +    fn try_clone(&self) -> io::Result<Self>;
> > +    /// Sets a read timeout.
> > +    fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()>;
>
> Do these take Option<Duration> because None means "Unset the timeout"? If
> so, please add comments saying so.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/tarpc/pull/27/files#r54045594>.
>